application/helpers/expressions/em_core_helper.php

```diff
@@ -2867,6 +2867,28 @@ function exprmgr_sumifop($args)
     return $result;
 }
 
+/**
+ * Validate a Gregorian date
+ * @see https://www.php.net/checkdate
+ * Check if all params are valid before send it to PHP checkdate to avoid PHP Warning
+ *
+ * @param mixed $month
+ * @param mixed $day
+ * @param mixed $year
+ * @return boolean
+ */
+function exprmgr_checkdate($month, $day, $year)
+{
+    if (
+        (!ctype_digit($month) && !is_int($month))
```

Comment:
Are ctype and is_int overlapping a bit? : `is_int('23') = bool(false)` is_int is a strict compare.
Not params in checkdate

If month is NOT an int **AND** not a digit : return false
**OR** If day is NOT an int **AND** not a digit : return false
**OR** If year is NOT an int **AND** not a digit : return false

Maybe a better test, more simple to understand ?

Reply:
> If month is NOT an int AND not a digit : return false

Shouldn't that be 
`If month is not an int OR not a digit : return false`

Reply:
exprmgr_checkdate('10', 10, 2022)

$month is NOT and integer 

`!ctype_digit($month) || !is_int($month)` === true

then `exprmgr_checkdate` return false.

It's not what we want

Reply:
I would thought that you would want to do some casting inside `exprmgr_checkdate`
What's the added value then? 

If returning false and no warning I guess it would be an unexpected result and confusing for the user, who may not know about type casting.

Reply:
`checkdate` is unusable with debug  > 2, but it's NOT a PHP issue …
It's potentially an expression issue … 

> If returning false and no warning I guess it would be an unexpected result and confusing for the user, who may not know about type casting.

Totally unrelated: simple user don't set debug . It's send PHP warning ONLY if debug set in config.php file 

And we have **exactly** the same fix for checkdate : https://github.com/LimeSurvey/LimeSurvey/blob/b634379e6f02565b6f2101ee3b0625d35110baec/application/helpers/expressions/em_core_helper.php#L2925

Reply:
Same system for

 https://github.com/LimeSurvey/LimeSurvey/blob/b634379e6f02565b6f2101ee3b0625d35110baec/application/helpers/expressions/em_core_helper.php#L2729

https://github.com/LimeSurvey/LimeSurvey/blob/b634379e6f02565b6f2101ee3b0625d35110baec/application/helpers/expressions/em_core_helper.php#L3095

https://github.com/LimeSurvey/LimeSurvey/blob/b634379e6f02565b6f2101ee3b0625d35110baec/application/helpers/expressions/em_core_helper.php#L3128

https://github.com/LimeSurvey/LimeSurvey/blob/b634379e6f02565b6f2101ee3b0625d35110baec/application/helpers/expressions/em_core_helper.php#L3155

etc …

Reply:
> If returning false and no warning I guess it would be an unexpected result and confusing for the user, who may not know about type casting.

with casting checkdate('12A', 12, 12) return true … but it's clearly false …

Reply:
> And we have **exactly** the same fix for `date` function :

is_numeric is different than is_int.
Also, they could do some casting, to transform a pure numeric string into an int. 
Not sure why they didn't.

Code is structurally OK, still, as we are adding the function from scratch, I believe it is a good opportunity to add more benefits than wrapping.
So, maybe the best is someone else to review it and approve it.

Reply:
But you don't want checkdate('12A', 12, 2022) == true ? Or checkdate(12.12 , 12, 2022) == true : right ?
Else : just remove this function … since it test nothing …

I really don't understand your point here …

Reply:
What about this? `checkdate('12', '12', '2022')` Should be true or false?

Reply:
true

Reply:
I think the current validations will make that to return false.

Reply:
![Capture d’écran du 2022-11-16 19-46-01](https://user-images.githubusercontent.com/1439428/202266623-85742614-51e4-410e-88f9-e52be8a4a799.png)


You test or you review ?

Reply:
Survey included in https://bugs.limesurvey.org/view.php?id=18401

Reply:
No, i didn't test it. 
I was reviewing the code and asked you.

You tested and it worked. Good.
Seems to be some internal casting which I didn't considered.

Reply:
Cast as int for clarity

https://github.com/LimeSurvey/LimeSurvey/pull/2675/commits/c76aa762c9988273f2291626513b3299ef303063